### ISSUE TITLE
Yale NVector#dot(nvector) causes segfault

### ISSUE THREAD

**Issue by translunar** [SciRuby/nmatrix#99]
For whatever reason, the following code produces an error:

```
n = NMatrix.new(:yale, [8,1], :int64) # also NVector
m = NMatrix.new(:yale, [1,8], :int64) # also NVector
n.dot(m)
```

This involves no change in `itype`, so it must be a problem with `symbmm`, the Yale matrix multiplication function.

Here is the `gdb` output:

```
Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_PROTECTION_FAILURE at address: 0x00007fff5fc11040
0x0000000103e76ac3 in nm::math::symbmm<unsigned short> (n=80, m=80, ia=0x7fff5fbfca60, ja=0x7fff5fbfca60, diaga=<value temporarily unavailable, due to optimizations>, ib=0x7fff5fbfca40, jb=0x7fff5fbfca40, diagb=true, ic=0x1035cb850, diagc=true) at util/math.h:1142
1142              mask[k] = i;
```

I'm unclear as to why this error hadn't arisen previously. In particular, this code is based on code supplied in #63, but scaled down in vector size so as not to trigger an itype-recast.

The same code works just fine when the matrix dimensions are `[8,2]` and `[2,8]` instead.

In other words, a new dot function is needed which handles Yale vectors.

It's also worth noting that the first Yale vector (`n`) is extremely inefficient. It uses 9 storage, more than dense! This is because it needs a diagonal entry for each row. We probably need to implement an alternative transposed storage form.
